### PR TITLE
Tasks/issue 32b

### LINKED ISSUE #32: Add Validation for Roles and points of contact
(Sections 3, 4, & 5) All required POCs are identified (System Owner, System ISSO, etc.)	

Simple Automation							

CSP can add POCs. Double-check to make sure all roles are listed in the Types of Users table.		

After talking with @brianrufgsa on 12/8/20, the rules are as follow:

1. Domain is simple many to many entity join: `metadata/role` M:M with `metadata/party` joined using to `metadata/responsible-party`
1. Referential Integrity Check: 
    - 	`metadata/responsible-party/@role-id` is in `metadata/role/@id`														
    - 	`metadata/responsible-party/party-uuid` is in `metadata/party/@uuid`	
1. Cardinality Check: FedRAMP specific and the role cardinality will be defined [here](https://github.com/18F/fedramp-automation/issues/32#issuecomment-740993340)
1. Roles and parties have no scope spanning outside of a document set: i.e. the baseline roles are defined for the authoring of the baseline, not as roles to be used in SSP's adhering to the baseline
1. UUID's are intended to be globally unique, but never changing for a given entity over time (there is no central store that maintains UUID & the associated entity over time, but it's an attempt to provide that).
1. UUID's may be scoped across related documents: 

> For the assessment plan and results, linking to the SSP will be unavoidable. The whole point is not to duplicate that content from the SSP into the SAP and SAR, rather to reference it back at the SSP.

### PR DESCRIPTION
**Pull request by mike-stern**
resolves #32

### CODE REVIEW

File: resources/validations/src/ssp.sch [hunk 1]
```diff
@@ -340,5 +340,16 @@
             >Response statement component remarks for <sch:value-of select="../../@statement-id"/> is too short with <sch:value-of select="$remarks-length"/> characters. It must be <sch:value-of select="$required-length"/> characters long.</sch:assert>
     </sch:rule>
 
-</sch:pattern>
-</sch:schema>
\ No newline at end of file
+ 
+    <sch:rule context="/o:system-security-plan/o:metadata">
+        <sch:let name="parties" value="o:party"/> 
+        <sch:let name="roles" value="o:role"/>
```
Comment (ohsh6o) [inline resources/validations/src/ssp.sch:346]: I think if you discuss `$extranenous-roles` you open us, logically, to discussing what `$required-roles` should be. You will not like where this is going!

Pure OSCAL specifies what is required in [the core Metaschema definition `o:allowed-values/@enum` in this part of the spec](https://github.com/usnistgov/OSCAL/blob/767a3ef90bc27a1a5c5a615b3de79a8139138030/src/metaschema/oscal_ssp_metaschema.xml#L101). This says `@allow-other = yes` so we are almost out of the clear.

In [the FedRAMP documentation](https://github.com/18F/fedramp-automation/blob/df177cbe77d36a0d0ee4b3b90059986b0b7a159c/documents/Guide_to_OSCAL-based_FedRAMP_System_Security_Plans_(SSP).pdf) on page 18 and the next few whereever a `role-id` comes up, but not in the current FedRAMP values, we see this one of those cases where FedRAMP will inevitably be more strict, they have minimum required ones as explained in the long-form documentation . These are not defined in the resolved catalog profile for the SSP, but in the OSCAL Metaschema or FedRAMP values. You are in a more fun case, you need to know those even if `@allow-other = no` because you are looking for the bare minimum required. This is a few new edge case for both of us.

If we rely on this code here, easy test, a SSP that does not define the responsible party in the metadata won't be there, but there will be no reported error. I assume this is the area where metaschema shines, so we need to figure this out, relying on work in #44 enhancements probably. Does that make sense?

Let's talk tomorrow. We can merge this and bring that in next for #32 before closing or leave this open.

**Review by ohsh6o — COMMENTED** [review]
Have some feedback we should discuss first. Solid approach, but you are hitting up against some of the same issues I am trying to sort out now in my branch! 😆

File: resources/validations/test/ssp.xspec [hunk 2]
```diff
@@ -842,4 +842,138 @@
             </x:scenario>
         </x:scenario>
     </x:scenario>
+
+    <x:scenario label="For an OSCAL FedRAMP SSP">
```
[the remark this answers is not in the document]
Comment (mike-stern) [reply resources/validations/test/ssp.xspec]: moved and pushed

File: resources/validations/src/ssp.sch (hunk 1 above)
Comment (mike-stern) [reply resources/validations/src/ssp.sch:346]: perhaps extraneous should be changed to undefined, given this is really a semantic referential integrity check scoped only to the SSP.

I like where you're going though, expanding this to also consider what OSCAL/Fedramp defines as known roles and/or required roles.

so to refine #32 a bit further, it sounds like we want to introduce the rules and/or reporting of: 

- validate that required roles are present--sounds like these required roles are not part of the OSCAL, but instead the user manual, which would require hard coding or using the "Fedramp Values file" as an input?
- report on the enumerated role values 
  - if `@allow-other = yes` potentially report on roles found that are not in the enum as an informational message, maybe even going so far as: "_did you mean to select a role from this defined list: definedRole1 definedRole2, ..._?"
  - if `@allow-other = no` assert on roles found that are not in the enum as an error message, again even going so far as: "_did you mean to select a role from this defined list: definedRole1 definedRole2, ..._?"

File: resources/validations/test/ssp.xspec (hunk 2 above)
[the remark this answers is not in the document]
Comment (ohsh6o) [reply resources/validations/test/ssp.xspec]: Thanks!

File: resources/validations/src/ssp.sch (hunk 1 above)
Comment (ohsh6o) [reply resources/validations/src/ssp.sch:346]: Ok given our conversation today [with Brian and Wendell in our Gitter channel](https://gitter.im/usnistgov-OSCAL/FedRAMP-10x-Schematron?at=60006f493567d7393d48f495), I guess we shall table this cardinality issue as we work with @brianrufgsa towards the new and improved FedRAMP extensions over simple FedRAMP values concept.

Comment (ohsh6o) [reply resources/validations/src/ssp.sch:346]: I created 18F/fedramp-automation#50 to track this later and link back to the origin of this refinement on these sorts of problems.
